libafl_bolts/src/simd.rs

```diff
@@ -0,0 +1,383 @@
+            let i = step * N;
+            let history = VectorType::new(hist[i..i + N].try_into().unwrap());
+            let items = VectorType::new(map[i..i + N].try_into().unwrap());
+
+            if items.max(history) != history {
+                interesting = true;
+                break;
+            }
+        }
+
+        if !interesting {
+            for j in (size - left)..size {
+                unsafe {
+                    let item = *map.get_unchecked(j);
+                    if item > *hist.get_unchecked(j) {
+                        interesting = true;
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    (interesting, novelties)
+}
+
+/// Coverage map insteresting naive implementation. Do not use it unless you have strong reasons to do.
+#[cfg(feature = "alloc")]
+#[must_use]
+pub fn covmap_is_interesting_naive(
+    hist: &[u8],
+    map: &[u8],
+    collect_novelties: bool,
+) -> (bool, Vec<usize>) {
+    let mut novelties = vec![];
+    let mut interesting = false;
+    let initial = 0;
+    if collect_novelties {
+        for (i, item) in map.iter().enumerate().filter(|(_, item)| **item != initial) {
+            let existing = unsafe { *hist.get_unchecked(i) };
+            let reduced = existing.max(*item);
+            if existing != reduced {
+                interesting = true;
+                novelties.push(i);
+            }
+        }
+    } else {
+        for (i, item) in map.iter().enumerate().filter(|(_, item)| **item != initial) {
+            let existing = unsafe { *hist.get_unchecked(i) };
+            let reduced = existing.max(*item);
+            if existing != reduced {
+                interesting = true;
+                break;
+            }
+        }
+    }
+
+    (interesting, novelties)
+}
+
+/// Standard coverage map instereting implementation. Use the fastest implementation by default.
+#[cfg(feature = "alloc")]
+#[allow(unused_variables)] // or we fail cargo doc
+#[must_use]
+#[rustversion::not(nightly)]
+pub fn std_covmap_is_interesting(
+    hist: &[u8],
+    map: &[u8],
+    collect_novelties: bool,
+) -> (bool, Vec<usize>) {
+    #[cfg(not(feature = "wide"))]
+    let ret = covmap_is_interesting_naive(hist, map, collect_novelties);
+
+    #[cfg(feature = "wide")]
+    let ret = covmap_is_interesting_u8x16(hist, map, collect_novelties);
+
+    ret
+}
+
+/// Standard coverage map instereting implementation on nightly. Use the fastest implementation by default.
+#[cfg(feature = "alloc")]
+#[allow(unused_variables)] // or we fail cargo doc
+#[must_use]
+#[rustversion::nightly]
+pub fn std_covmap_is_interesting(
+    hist: &[u8],
+    map: &[u8],
+    collect_novelties: bool,
+) -> (bool, Vec<usize>) {
+    #[cfg(not(feature = "wide"))]
+    let ret = covmap_is_interesting_stdsimd(hist, map, collect_novelties);
+
+    #[cfg(feature = "wide")]
+    let ret = covmap_is_interesting_u8x16(hist, map, collect_novelties);
```

Comment:
is "u8x16" default? it's not 32?

Reply:
ok "slightly slower than u8x16"

Reply:
Correct. The rationale is that `u8x16` is fast enough in most cases while `u8x32` might not available on some older cpu models (I think before Haswell at least). In addition, on x86 I observed that `u8x16` is faster while on aarch64 `u8x32` is faster by ~10%. Therefore, I think defaulting to `u8x16` is pretty enough, or I can also specify different default methods for different architectures.